bitcoin/src/pow.rs

```diff
@@ -968,44 +969,71 @@ where
     let adjustment_interval = consensus_params.difficulty_adjustment_interval();
     let pow_target_timespan = consensus_params.pow_target_timespan;
     let no_pow_retargeting = consensus_params.no_pow_retargeting;
-    let last_adjustment_height =
-        current_height.to_consensus_u32() as u64 - (adjustment_interval - 1);
+    let allow_min_difficulty_blocks = consensus_params.allow_min_difficulty_blocks;
+    let pow_target_spacing = consensus_params.pow_target_spacing as u32;
+
+    let current_height = current_height.to_consensus_u32() as u64;
+    let current_time = current_header.time;
+
+    let last_adjustment_height = current_height - (adjustment_interval - 1);
     let last_adjustment_block = get_block_header_by_height(last_adjustment_height as u32)?;
     let last_adjustment_time = last_adjustment_block.time;
-    let current_height: u64 = current_height.to_consensus_u32() as u64;
 
-    // TODO(vincenzopalazzo): in the testnet case there is some rule
-    // where if there is no block for a long time the
-    // difficulty drops to 1, or something.
-    if ((current_height) % adjustment_interval.min(1)) != 0 {
-        return Ok(last_adjustment_block.target().to_compact_lossy().to_consensus());
+    if (current_height + 1) % adjustment_interval != 0 {
+        if allow_min_difficulty_blocks {
+            let pow_limit =
+                consensus_params.pow_limit.to_target().to_compact_lossy().to_consensus();
+            if new_block_timestamp > current_header.time + pow_target_spacing * 2 {
+                return Ok(pow_limit);
+            } else {
+                use hashes::Hash;
+                let mut header = current_header;
+                let mut height = current_height;
+                while header.prev_blockhash != BlockHash::all_zeros()
+                    && height % adjustment_interval != 0
+                    && header.bits.to_consensus() == pow_limit
+                {
+                    height -= 1;
+                    header = get_block_header_by_height(height as u32)?;
+                }
+                return Ok(header.bits.to_consensus());
+            }
+        }
+        return Ok(current_header.bits.to_consensus());
     }
 
     if no_pow_retargeting {
-        return Ok(last_adjustment_block.bits.to_consensus());
+        return Ok(current_header.bits.to_consensus());
     }
 
     let actual_timespan = (current_time - last_adjustment_time) as u64;
-    let adjusted_timespan = actual_timespan;
+    let mut adjusted_timespan = actual_timespan;
 
     if actual_timespan < pow_target_timespan / 4 {
-        return Ok((pow_target_timespan / 4) as u32);
+        adjusted_timespan = pow_target_timespan / 4;
     }
     if actual_timespan > pow_target_timespan * 4 {
-        return Ok((pow_target_timespan / 4) as u32);
+        adjusted_timespan = pow_target_timespan * 4;
     }
```

Comment:
we can return the value here? as this review suggests?  https://github.com/rust-bitcoin/rust-bitcoin/pull/2090#discussion_r1335180944

Or this change the logic?

Reply:
`return Ok((pow_target_timespan / 4) as u32);` doesn't make any sense.

You are returning `302400` for every single call, regardless of difficulty.

`pow_target_timespan` is "The number of seconds in 2 weeks" (for bitcoin mainnet) which is always `1209600`

Dividing by 4 is always `302400`

This calculation has nothing to do with Target, or Work, or difficulty, or PoW.

Reply:
Returning early doesn't make much sense either, since you will still need to do all the calculations in the lines below.

Copy pasting those 18 lines for both of these if statements doesn't make much sense.